### PR TITLE
fix: DeepPickFromPath<> to flatten Arrays of Models referenced in a join table

### PR DESCRIPTION
**Pull request by zxl629**
*Related issues:*
- https://github.com/aws-amplify/amplify-category-api/issues/3062

*Description of changes:*

This change adds a defined-behavior integration test for custom selection set. The test creates a model that serves as "join table" with two one-to-many relationships between two entities. The test then uses a query on the "join table" model with a custom selection set to get one of the entities. It will fail since that entity has an array field, and `DeepPickFromPath()` is unable to flatten that array, which causes errors in type intersections.

The change then fixes the bug by adding a check in `DeepPickFromPath()` for custom type in a model's fields. If the field is a custom type, the function will flatten the array fields inside that data model as well. 

*Example of why we need this:*

If customers have two data models `Part` and `Customer`, and a model `CustomerPart` that serves as a "join table":

```ts
Customer: a.model({
          name: a.string().required(),
          verbiageReqReq: a.ref('Verbiage').required().array().required(),
          verbiageOptReq: a.ref('Verbiage').array().required(),
          verbiageReqOpt: a.ref('Verbiage').required().array(),
          verbiageOptOpt: a.ref('Verbiage').array(),
          orders: a.hasMany('CustomerPart', 'customerId'),
        }),
CustomerPart: a.model({
          customer: a.belongsTo('Customer', 'customerId'),
          customerId: a.string(),
          experience: a.string(),
          part: a.belongsTo('Part', 'partId'),
          partId: a.string(),
        }),
Part: a.model({
          name: a.string().required(),
          orders: a.hasMany('CustomerPart', 'partId'),
        }),
```
when they try to query the customer's data from `CustomerPart` using a selection set:
```ts
const { data } = await client.models.CustomerPart.list({
          selectionSet: [
            'customer.*',
            'customer.verbiageReqReq.*',
            'customer.verbiageOptReq.*',
            'customer.verbiageReqOpt.*',
            'customer.verbiageOptOpt.*',
          ],
        });
```
they will get wrong types in their results:
```ts
const cps = data[0].customer;
        const verbiageReqReq = cps.verbiageReqReq; // should be Verbiage[] and it is
        const verbiageOptReq = cps.verbiageOptReq; // should be (Verbiage | null)[], but is (never[] | null)[]
        const verbiageReqOpt = cps.verbiageReqOpt; // should be Verbiage[] | null, but is never[][] | null
        const verbiageOptOpt = cps.verbiageOptOpt; // should be (Verbiage | null)[] | null, but is (never[] | null)[] | null
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by stocaaro — APPROVED** [review]
This change LGTM. Glad to have the added testing to cover this usecase. I have one question about test organization re: type tests, but am good to merge if we're happy with how that is situated.

File: packages/integration-tests/__tests__/defined-behavior/3-exhaustive/custom-selection-set.ts [hunk 1]
```diff
@@ -1107,4 +1107,159 @@ describe('Custom selection set edge cases', () => {
...
+        type ActualType = typeof cus;
+        type _test = Expect<Equal<ActualType, ExpectedType>>;
+      });
+
+      test('.array().required()', async () => {
+        const { data } = await mockedOperation();
+        const cus = data[0].customer.verbiageOptReq;
+        type ExpectedType = (Verbiage | null)[];
+        type ActualType = typeof cus;
+        type _test = Expect<Equal<ActualType, ExpectedType>>;
```
[the remark this answers is not in the document]
Comment (zxl629) [reply packages/integration-tests/__tests__/defined-behavior/3-exhaustive/custom-selection-set.ts:1245]: I think you are right, moving this test to a type test file seems more appropriate.

[the remark this answers is not in the document]
Comment (iartemiev) [reply packages/integration-tests/__tests__/defined-behavior/3-exhaustive/custom-selection-set.ts:1245]: For the `defined-behavior/*` tests we do not need to separate type assertions and runtime assertions. They should be colocated.

`*.test-d.ts`distinction was only necessary for the in-package unit/fn tests

**Review by iartemiev — APPROVED** [review]
